lib/puppet/rest/routes.rb

```diff
@@ -12,79 +16,134 @@ def self.ca
+    # @return [String] the PEM-encoded crl
+    def self.get_crls(name, ssl_context)
+      ca.with_base_url(Puppet::Network::Resolver.new) do |url|
         header = { 'Accept' => 'text/plain', 'Accept-Encoding' => ACCEPT_ENCODING }
         url.path += "certificate_revocation_list/#{name}"
-        client.get(url, header: header) do |chunk|
-          block.call(chunk)
+
+        use_ssl = url.is_a? URI::HTTPS
+
+        # Deeper levels of the code assume that if we have any number of
+        # certificate related files, we have all of the certificate related
+        # files. Unfortunately, this causes us to get stuck in an infinite loop,
+        # so we have to hard code `verify_mode=false` so we don't attempt to use
+        # files that do not exist yet in order to download those files.
+        #
+        # This is related to PUP-9094. We won't have so many issues with this
+        # once we are using the httpclient gem to handle this work. We were
+        # unable to get this work completed in time for Puppet 6.0.0, so we had
+        # to switch back to using Puppet::Network::HttpPool, which has
+        # unfortunate limitations (i.e., an all or nothing approach to cert
+        # verification).
+        verify_mode = false
+
+        client = Puppet::Network::HttpPool.http_instance(url.host, url.port, use_ssl, verify_mode)
+
+        response = client.get(url.request_uri, header)
+        unless response.code.to_i == 200
+          raise Puppet::Rest::ResponseError.new(response.message, response)
         end
+
         Puppet.debug _("Downloaded certificate revocation list for %{name} from %{server}") % { name: name, server: ca.server }
+
+        uncompress_body(response)
       end
     end
 
-    # Make an HTTP request to send the named CSR, using the given
-    # HTTP client.
-    # @param [Puppet::Rest::Client] client the HTTP client to use to make the request
+    # Make an HTTP request to send the named CSR
     # @param [String] csr_pem the contents of the CSR to sent to the CA
     # @param [String] name the name of the host whose CSR is being submitted
+    # @param [Puppet::Rest::SSLContext] ssl_context the ssl content to use when making the request
     # @rasies [Puppet::Rest::ResponseError] if the response status is not OK
-    def self.put_certificate_request(client, csr_pem, name)
-      ca.with_base_url(client.dns_resolver) do |url|
+    def self.put_certificate_request(csr_pem, name, ssl_context)
+      ca.with_base_url(Puppet::Network::Resolver.new) do |url|
         header = { 'Accept' => 'text/plain',
                    'Accept-Encoding' => ACCEPT_ENCODING,
                    'Content-Type' => 'text/plain' }
         url.path += "certificate_request/#{name}"
-        response = client.put(url, body: csr_pem, header: header)
-        if response.ok?
+
+        use_ssl = url.is_a? URI::HTTPS
+        verify_mode = ssl_context.verify_mode == OpenSSL::SSL::VERIFY_PEER
```

Comment:
Why are we able to use VERIFY_PEER here? I feel since we don't have a host cert yet this should break? Unless it's just falling back automatically?